packages/python/plotly/plotly/express/_core.py

```diff
@@ -1341,46 +1420,104 @@ def build_dataframe(args, constructor):
 
     # Cast data_frame argument to DataFrame (it could be a numpy array, dict etc.)
     df_provided = args["data_frame"] is not None
-    needs_interchanging = False
-    if df_provided and not isinstance(args["data_frame"], pd.DataFrame):
-        if hasattr(args["data_frame"], "__dataframe__") and version.parse(
-            pd.__version__
-        ) >= version.parse("2.0.2"):
-            import pandas.api.interchange
-
-            df_not_pandas = args["data_frame"]
-            args["data_frame"] = df_not_pandas.__dataframe__()
-            # According interchange protocol: `def column_names(self) -> Iterable[str]:`
-            # so this function can return for example a generator.
-            # The easiest way is to convert `columns` to `pandas.Index` so that the
-            # type is similar to the types in other code branches.
-            columns = pd.Index(args["data_frame"].column_names())
-            needs_interchanging = True
-        elif hasattr(args["data_frame"], "to_pandas"):
-            args["data_frame"] = args["data_frame"].to_pandas()
+    is_pd_like = False
+    if df_provided:
```

Comment:
How about rewriting this as
```python
        if nw.dependencies.is_pandas_like_dataframe(args["data_frame"]):

            columns = args["data_frame"].columns  # This can be multi index
            args["data_frame"] = nw.from_native(args['data_frame'], eager_only=True)
            is_pd_like = True

        elif nw.dependencies.is_pandas_like_series(args["data_frame"]):

            args["data_frame"] = nw.from_native(args['data_frame'], series_only=True).to_frame()
            columns = args["data_frame"].columns
            is_pd_like = True

        elif isinstance(data_frame := nw.from_native(args['data_frame'], eager_only=True, strict=False), nw.DataFrame):
            args["data_frame"] = data_frame
            columns = args['data_frame'].columns

        elif isinstance(series := nw.from_native(args['data_frame'], series_only=True, strict=False), nw.Series):
            args["data_frame"] = series.to_frame()
            columns = args['data_frame'].columns
```
?

Then, if/when we support other eager dataframe in Narwhals, Plotly would get support for them for free

Reply:
That's wonderful! Let me fix that!